### PR TITLE
Initial commit of `Ptr` type

### PR DESCRIPTION
**Pull request by joshlf**
`Ptr` is like `NonNull`, but has many restrictions which make it so that using a `Ptr` in unsafe code requires much simpler soundness proofs. Notably, it supports a `try_cast_into<U>` method where `U: ?Sized + KnownLayout`, which is a building block of `TryFromBytes`.

### CODE REVIEW

File: src/lib.rs
Comment (jswrenn) [inline src/lib.rs:612]: Okay, I've *finally* thought of something better than "no_affix":
```suggestion
    fn try_cast_from_no_excess(bytes: NonNull<[u8]>) -> Option<NonNull<Self>> {
```

Comment (joshlf) [reply src/lib.rs]: I still feel like `prefix_suffix`, though a bit more verbose, is the most straightforward and descriptive.

Comment (jswrenn) [reply src/lib.rs]: Maybe? I'm a little concerned that prefix and suffix are getting a bit overloaded. A prefix cast isn't one with no prefix, and a suffix cast isn't one with no suffix. If anything, this is a cast in which prefix == suffix.

File: src/util.rs
Comment (jswrenn) [inline src/util.rs:52]: In cases where a marker field can be given a more specific name, I prefer to use that name:
```suggestion
    _lifetime: PhantomData<&'a ()>,
```

File: src/util.rs
Comment (jswrenn) [inline src/util.rs:40]: Like `NonNull`, we should probably document the variance of the parameters. Here, both `'a` and `T` are covariant.

**Review by jswrenn — CHANGES_REQUESTED** [review]

File: src/util.rs
Comment (joshlf) [reply src/util.rs:463]: Discussed offline; we're going to just disable symbolic alignment checking because the alternative introduces more complexity and CI test time than it's worth for a marginal benefit.

File: src/lib.rs
Comment (joshlf) [reply src/lib.rs]: Discussed offline; we're going with `try_cast_from_no_leftover`.

File: src/util.rs
Comment (joshlf) [reply src/util.rs]: Done.

File: src/util.rs
Comment (joshlf) [reply src/util.rs:44]: Done.

File: src/util.rs
Comment (jswrenn) [inline src/util.rs:160]: Considering that "SAFETY" is used to introduce a comment that explains why code is actually safe, it's a bit odd that "PANICS" is introducing a comment about why something *doesn't* panic. I think you could just append "and thus `validate_cast_and_convert_metadata` won't panic.

File: src/util.rs
Comment (jswrenn) [inline src/util.rs:273]: It's regrettable that we don't yet have unsafe fields, or unsafe implicit instructors. What do you think about renaming the `ptr` field to `unsafe_ptr`, just to hammer the point home that this field has safety invariants at its use-sites?

File: src/util.rs
Comment (jswrenn) [inline src/util.rs:395]: This is subtle. Could you cite the reference that fat pointer conversions preserve length?

Comment (joshlf) [reply src/util.rs:395]: So fun discovery; there is no such guarantee as of today. I've put up a PR to add one: https://github.com/rust-lang/reference/pull/1417. We'll need to block this PR until that lands.

File: src/macros.rs
Comment (joshlf) [reply src/macros.rs:215]: We don't; I've added a TODO.

File: src/util.rs
Comment (joshlf) [reply src/util.rs]: Done.

File: src/util.rs
Comment (joshlf) [reply src/util.rs:396]: Another option would be to split this PR so that all of the bits that transitively rely on this method are in a follow-up PR. That would include the `try_from` method, so it'd be a good chunk of this PR that we'd need to save for a follow-up.

Comment (joshlf) [reply src/util.rs]: Discussed offline. I've split this and transitive dependencies into https://github.com/google/zerocopy/pull/498.

File: src/util.rs
Comment (joshlf) [reply src/util.rs:143]: I'm worried that if we do that, then for consistency's sake we'll need to do it on almost every field in the crate. I feel like if you're working on this crate, you've gotta already be aware that most things have safety invariants (and certainly code reviewers know that), so the risk of missing an invariant is low. Another angle is that the sort of person who wouldn't know to check whether `ptr` has a safety invariant is the same sort of person who wouldn't get the hint from the name `unsafe_ptr`.

Comment (joshlf) [reply src/util.rs]: Yeah fair enough; I've moved it into a submodule.

**Review by jswrenn — APPROVED** [review]

**Pull request merged**
